File: Data Structures & Algorithms/interleaving-string/submission-3.py
class Solution:
   
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        self.dp = {}
        if (len(s1) + len(s2) != len(s3)):
            return False;
        return self.dfs(0,0,0,s1,s2,s3)
    
    def dfs(self,i: int, j:int,k:int ,s1: str, s2: str, s3: str)->bool:
        if (k == len(s3)):
            return (i == len(s1) and j == len(s2))
        if ((i,j) in self.dp) :
            return self.dp[(i,j)]
        res = False
        if (i < len(s1) and s1[i] == s3[k]):
            res = self.dfs(i+1,j,k+1,s1,s2,s3)
        if (not res and j < len(s2) and s2[j] == s3[k]):
            res = self.dfs(i,j+1,k+1,s1,s2,s3)
        self.dp[(i,j)] = res
        return res

File: Data Structures & Algorithms/surrounded-regions/submission-3.java
class Solution {
    public void solve(char[][] board) {
        int rows = board.length;
        int cols = board[0].length;
        if (rows == 0) return;
        
        // Step 1: Mark all 'O's connected to the borders
        for (int i = 0; i < rows; i++) {
            dfs(i, 0, board);
            dfs(i, cols - 1, board);
        }
        for (int j = 0; j < cols; j++) {
            dfs(0, j, board);
            dfs(rows - 1, j, board);
        }

        // Step 2: Traverse the board, flip 'O' to 'X' and 'T' back to 'O'
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (board[i][j] == 'O') board[i][j] = 'X';
                else if (board[i][j] == 'T') board[i][j] = 'O';
            }
        }
    }

    private void dfs(int r, int c, char[][] board) {
        if (r < 0 || c < 0 || r >= board.length || c >= board[0].length || board[r][c] != 'O') return;
        board[r][c] = 'T'; // Mark as temporary to indicate it's not surrounded
        dfs(r + 1, c, board);
        dfs(r - 1, c, board);
        dfs(r, c + 1, board);
        dfs(r, c - 1, board);
    }
}